js/models/messages.js

```diff
@@ -1935,105 +1977,212 @@
+              );
+            }
 
-              const membersMissing =
-                _.difference(
-                  conversation.get('members'),
-                  initialMessage.group.members
-                ).length > 0;
+            const membersMissing =
+              _.difference(
+                conversation.get('members'),
+                initialMessage.group.members
+              ).length > 0;
 
-              if (membersMissing) {
-                window.log.warn('Non-admin attempts to remove group members');
-              }
+            if (membersMissing) {
+              window.log.warn('Non-admin attempts to remove group members');
+            }
 
-              const messageAllowed = !nameChanged && !membersMissing;
+            const messageAllowed = !nameChanged && !membersMissing;
 
-              if (!messageAllowed) {
-                confirm();
-                return null;
-              }
+            // Returning true drops the message
+            if (!messageAllowed) {
+              confirm();
+              return true;
             }
           }
-          // For every member, see if we need to establish a session:
-          initialMessage.group.members.forEach(memberPubKey => {
-            const haveSession = _.some(
-              textsecure.storage.protocol.sessions,
-              s => s.number === memberPubKey
-            );
+        }
+        // send a session request for all the members we do not have a session with
+        window.libloki.api.sendSessionRequestsToMembers(
+          initialMessage.group.members
+        );
+      } else if (newGroup) {
+        // We have an unknown group, we should request info from the sender
+        textsecure.messaging.requestGroupInfo(conversationId, [primarySource]);
+      }
+      return false;
+    },
 
-            const ourPubKey = textsecure.storage.user.getNumber();
-            if (!haveSession && memberPubKey !== ourPubKey) {
-              ConversationController.getOrCreateAndWait(
-                memberPubKey,
-                'private'
-              ).then(() => {
-                textsecure.messaging.sendMessageToNumber(
-                  memberPubKey,
-                  '(If you see this message, you must be using an out-of-date client)',
-                  [],
-                  undefined,
-                  [],
-                  Date.now(),
-                  undefined,
-                  undefined,
-                  { messageType: 'friend-request', sessionRequest: true }
-                );
-              });
-            }
-          });
-        } else if (newGroup) {
-          // We have an unknown group, we should request info from the sender
-          textsecure.messaging.requestGroupInfo(conversationId, [
-            primarySource,
-          ]);
+    async handleSessionRequest(source, primarySource, confirm) {
+      // Check if the contact is a member in one of our private groups:
```

Comment:
a session request can only come from a user which is present in one of our closed groups